src/network/operator.rs

```diff
@@ -0,0 +1,402 @@
+                if let Some(block_template) = self.block_template.read().await.clone() {
+                    // Ensure this prover exists in the list first, and retrieve their share difficulty.
+                    let share_difficulty = self
+                        .provers
+                        .write()
+                        .await
+                        .entry(address)
+                        .or_insert((Instant::now(), BASE_SHARE_DIFFICULTY, 0))
+                        .1;
+
+                    // Route a `PoolRequest` to the peer.
+                    let message = Message::PoolRequest(share_difficulty, Data::Object(block_template));
+                    if let Err(error) = self.peers_router.send(PeersRequest::MessageSend(peer_ip, message)).await {
+                        warn!("[PoolRequest] {}", error);
+                    }
+                } else {
+                    warn!("[PoolRegister] No current block template exists");
+                }
+            }
+            OperatorRequest::PoolResponse(peer_ip, block_header, prover_address) => {
+                if let Some(block_template) = self.block_template.read().await.clone() {
+                    // Check that the block is relevant.
+                    if self.ledger_reader.latest_block_height().saturating_add(1) != block_header.height() {
+                        warn!("[PoolResponse] Peer {} sent a stale candidate block.", peer_ip);
+                        return;
+                    }
+
+                    // Ensure the nonce hasn't been seen before.
+                    if self.known_nonces.read().await.contains(&block_header.nonce()) {
+                        warn!("[PoolResponse] Peer {} sent a duplicate share", peer_ip);
+                        // TODO (julesdesmit): punish?
+                        return;
+                    }
+
+                    // Update known nonces.
+                    self.known_nonces.write().await.insert(block_header.nonce());
+
+                    // Reconstruct the block.
+                    let previous_block_hash = block_template.previous_block_hash();
+                    let transactions = block_template.transactions().clone();
+
+                    if let Ok(block) = Block::from_unchecked(previous_block_hash, block_header.clone(), transactions) {
+                        // Ensure the proof is valid.
+                        let proof = match block.header().proof() {
+                            Some(proof) => proof,
+                            None => {
+                                warn!("[PoolResponse] proof is missing on header");
+                                return;
+                            }
+                        };
+
+                        // NOTE (julesdesmit): Unwraps are here for brevity's sake, and since we do
+                        // it exactly the same in snarkVM, I don't see why we shouldn't use them here.
+                        let inputs = vec![
```

Comment:
nit: since we only borrow `inputs` later on, maybe this allocation can be avoided?

Reply:
The function currently expects a `&Vec` to be passed to it, so I can't circumvent it for now.

Reply:
it can be done later; in general there's no reason to use `&Vec<T>` instead of `&[T]`, so it could be done in a pass targeting such APIs